lib/ops/utils.go

```diff
@@ -300,3 +320,48 @@ func GetExpandOperation(backend storage.Backend) (*storage.SiteOperation, error)
 	}
 	return nil, trace.NotFound("expand operation not found")
 }
+
+// MatchByType returns an OperationMatcher to match operations by type
+func MatchByType(opType string) OperationMatcher {
+	return func(op SiteOperation) bool {
+		return op.Type == opType
+	}
+}
+
+// OperationMatcher is a function type that matches the given operation
+type OperationMatcher func(SiteOperation) bool
+
+// GetExistingDockerConfig returns existing cluster Docker configuration
+func GetExistingDockerConfig(
+	clusterKey SiteKey,
+	operator Operator,
+	manifestDocker schema.Docker,
+) (config *storage.DockerConfig, err error) {
+	config = &storage.DockerConfig{
+		StorageDriver: manifestDocker.StorageDriver,
+		Args:          manifestDocker.Args,
+	}
+
+	installOperation, err := GetCompletedInstallOperation(clusterKey, operator)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	installDocker := installOperation.InstallExpand.Vars.System.Docker
+	if installDocker.StorageDriver != "" {
+		config.StorageDriver = installDocker.StorageDriver
+	}
+	config.Args = append(config.Args, installDocker.Args...)
+
+	updateOperation, err := GetLastCompletedUpdateOperation(clusterKey, operator)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if updateOperation != nil &&
```

Comment:
Hm, I wonder if that's a reliable approach - to dynamically construct "docker" config by querying info from past operations every time. As opposed to, say, saving up-to-date info in ClusterState or something.

Reply:
Yes, I already changed this. Bigger reason to keep the state is that after the update the Docker configuration of the previous application is replaced (as it is part of the application manifest) and it will not be possible to deduce the current Docker config otherwise.